src/fluree/db/query/optimize.cljc

```diff
@@ -36,3 +44,184 @@
                 nxt
                 (reduced nxt)))
             (map compare-component a' b'))))
+
+(defn optimizable-pattern?
+  "Returns true for patterns we know how to optimize.
+   Any pattern not recognized here is treated as an optimization boundary."
+  [pattern]
+  (let [ptype (where/pattern-type pattern)]
+    (#{:tuple :class :id} ptype)))
+
+(defn get-iri
+  "Extract IRI from a matched component"
+  [component]
+  (where/get-iri component))
+
+(defn encode-iri-to-sid
+  "Convert IRI to SID using database namespace table"
+  [db iri]
+  (when iri
+    (iri/encode-iri db iri)))
+
+(defn get-property-count
+  "Get count for a property from stats"
+  [stats sid]
+  (get-in stats [:properties sid :count]))
+
+(defn get-class-count
+  "Get count for a class from stats"
+  [stats sid]
+  (get-in stats [:classes sid :count]))
+
+(defn matched?
+  "Check if a component is matched (has a bound value, IRI, or SID)"
+  [component]
+  (where/matched? component))
+
+(defn unmatched?
+  "Check if a component is unmatched (variable)"
+  [component]
+  (not (matched? component)))
+
+(defn all-matched?
+  "Check if all components are matched"
+  [[s p o]]
+  (and (matched? s) (matched? p) (matched? o)))
+
+(defn calculate-selectivity
+  "Calculate selectivity score for a pattern.
+   Lower score = more selective = execute first."
+  [db stats pattern]
+  (let [pattern-type (where/pattern-type pattern)
+        pattern-data (where/pattern-data pattern)]
+
+    (cond
+      (or (nil? stats) (empty? stats))
+      nil
```

Comment:
If we don't have the information to calculate the score it feels like we should use `default-selectivity` instead of `nil`, especially if we're going to be sorting on this later.

Reply:
This is no longer the case in the upstream branch, this was more a placeholder for this PR which was a base for the next HLL optimization one.